content/blog/2018-06-07-when-to-use-derived-state.md

```diff
@@ -0,0 +1,384 @@
+
+  render() {
+    const filteredList = this.props.list.filter(
+      item => item.text.includes(this.state.filterText)
+    )
+
+    return (
+      <Fragment>
+        <input onChange={this.handleChange} value={this.state.filterText} />
+        <ul>{filteredList.map(item => <li key={item.id}>{item.text}</li>)}</ul>
+      </Fragment>
+    );
+  }
+
+  handleChange = event => {
+    this.setState({ filterText: event.target.value });
+  };
+}
+```
+
+The above approach is much cleaner and simpler than the derived state version, but it has some limitations. Filtering may be slow for large lists, and `PureComponent` won't prevent re-renders if props other than `list` change. In that case, we could add a memoization helper to avoid unnecessarily re-filtering our list:
+
+```js
+import memoize from "memoize-one";
+
+class Example extends Component {
+  // State only needs to hold the current filter text value:
+  state = {
+    filterText: ""
+  };
+
+  // Re-run the filter whenever the list array or filter text change:
+  filter = memoize(
+    (list, filterText) => list.filter(
+      item => item.text.includes(filterText)
+    )
+  );
+
+  render() {
+    const filteredList = this.filter(this.props.list, this.state.filterText);
+
+    return (
+      <Fragment>
+        <input onChange={this.handleChange} value={this.state.filterText} />
+        <ul>{filteredList.map(item => <li key={item.id}>{item.text}</li>)}</ul>
+      </Fragment>
+    );
+  }
+
+  handleChange = event => {
+    this.setState({ filterText: event.target.value });
+  };
+}
+```
+
+This is much simpler and should perform just as well as the previous version!
+
+There are a couple of constraints to consider when using memoization though:
+
+1. In most cases, you'll want to **attach the memoized function to a component instance**. This prevents multiple instances of a component from resetting each other's memoized keys.
+1. Typically you'll want to use a memoization helper with a **limited cache size** in order to prevent memory from "leaking" over time. (In the example above, we used `memoize-one` because it only caches the most recent argument and result.)
+1. Memoization will not work properly if `props.items` is recreated each time the parent component renders. Hopefully, this should not be the case for large arrays. If it were, you'd want to address that problem first! (This limitation also applies to the `getDerivedStateFromProps` version above.)
+
+## In closing
+
+The examples above are intentionally simplified in order to highlight specific coding patterns. In real world applications, components often contain a mix of controlled and uncontrolled behaviors. This is okay! Just be careful to ensure that each behavior has a clear source of truth in order to avoid the anti-patterns mentioned above.
+
+If you have a use case that you think falls outside of these patterns, please share it with us on [GitHub](https://github.com/reactjs/reactjs.org/issues/new) or Twitter!
```

Comment:
One thing we haven't covered at all is Redux. In that case it's not clear what the recommended solution is (you can't lift draft state up without making it slow).

Reply:
Not sure I want to cover Redux in this post. I haven't used Redux in a long time, and I haven't thought about that use case much as a result. Do you think it's necessary?

Reply:
I think it's necessary (it seems like one of the most popular reasons why people even *want* to "mirror" state—the committed one lives in Redux, and the draft is in local state). Judging by the thread, Redux folks also seem to often be unaware of this problem because PureComponent obscures it.

I don't think we need to mention it by name or dedicate a section, but just casually mentioning in an appropriate section that it *also* serves as a solution when the source of truth is outside React would help.

Reply:
I'm not sure which section that would be by the way.

Reply:
Okay. Maybe something like this...

<img width="737" alt="screen shot 2018-06-05 at 8 54 51 pm" src="https://user-images.githubusercontent.com/29597/41009989-b950b4d2-6902-11e8-9215-d053d5b9515c.png">

Reply:
Redux users tend to want to reset it though.. :-(

Reply:
The rest of the blog post covers that. I don't think we need to go into great depth about Redux, specifically, in this section.

Reply:
Yeah upon re-reading I think this is sufficient

Reply:
Nice! 😄